logr.go

```diff
@@ -395,51 +395,50 @@ func (l Logger) WithContext(ctx context.Context) Logger {
 	return l
 }
 
-// WithContextValues extends the list of context keys and the name for them. When
-// given a context through WithContext later, the Logger will extract the
-// values for these keys and log them as additional key/value pairs
+// WithContextValues extends the list of context handlers. When
+// given a context through WithContext later, the Logger will invoke
+// each handler and log the key/value pairs that they return as
 // as if they had been passed to WithValues.
-func (l Logger) WithContextValues(keys ...ContextKey) Logger {
+func (l Logger) WithContextValues(handlers ...ContextHandler) Logger {
 	if l.sink == discardLogSinkInstance {
 		return l
 	}
 
-	if l.contextKeys == nil {
+	if l.contextHandlers == nil {
 		// Copy the parameters to avoid surprises when the caller changes the content
 		// of the parameter slice later.
-		cp := make([]ContextKey, len(keys))
-		copy(cp, keys)
-		l.contextKeys = &cp
+		cp := make([]ContextHandler, len(handlers))
+		copy(cp, handlers)
+		l.contextHandlers = &cp
 	} else {
 		// We must create a new slice because the existing one is shared between
 		// Logger instances.
-		cp := make([]ContextKey, len(*l.contextKeys)+len(keys))
-		copy(cp, *l.contextKeys)
-		copy(cp[len(*l.contextKeys):], keys)
-		l.contextKeys = &cp
+		cp := make([]ContextHandler, len(*l.contextHandlers)+len(handlers))
+		copy(cp, *l.contextHandlers)
+		copy(cp[len(*l.contextHandlers):], handlers)
+		l.contextHandlers = &cp
 	}
 	return l
 }
 
 func (l Logger) valuesFromContext() []interface{} {
-	if l.ctx == nil || l.contextKeys == nil {
+	if l.ctx == nil || l.contextHandlers == nil {
 		return nil
 	}
-	keysAndValues := make([]interface{}, 0, 2*len(*l.contextKeys))
-	for _, key := range *l.contextKeys {
-		if value := l.ctx.Value(key.Key); value != nil {
-			keysAndValues = append(keysAndValues, key.Name, value)
+	// We don't really know how large this has to be. Let's
+	// assume that each handler returns at most two key/value
+	// pairs, of which each needs two entries.
+	keysAndValues := make([]interface{}, 0, 4*len(*l.contextHandlers))
```

Comment:
There's almost no cost to over-estimating (within reason) so I'd go for like 64 instead of 4

Reply:
I had my math wrong: I meant `2 * (len(*l.contextHandlers) + 1)` = "one value might be an error, which then adds one more extra field with two values". But I agree, we can make that larger, just in case.

Reply:
Never mind, I had the context (sic!) wrong. Pure multiplications is right here.